// jfrog-client-js/dist/src/Xray/XraySystemClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XraySystemClient = void 0;
class XraySystemClient {
    constructor(httpClient, logger) {
        this.httpClient = httpClient;
        this.logger = logger;
        this.pingEndpoint = '/api/v1/system/ping';
        this.versionEndpoint = '/api/v1/system/version';
    }
    async ping() {
        this.logger.debug('Sending ping request...');
        const requestParams = {
            url: this.pingEndpoint,
            method: 'GET',
            timeout: 2000,
        };
        try {
            return await this.httpClient.doRequest(requestParams);
        }
        catch (error) {
            return false;
        }
    }
    async version() {
        this.logger.debug('Sending version request...');
        const requestParams = {
            url: this.versionEndpoint,
            method: 'GET',
        };
        return await this.httpClient.doAuthRequest(requestParams);
    }
}
exports.XraySystemClient = XraySystemClient;
