// jfrog-client-js/dist/src/Xray/XrayClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XrayClient = void 0;
const HttpClient_1 = require("../HttpClient");
const XraySystemClient_1 = require("./XraySystemClient");
const XrayLogger_1 = require("./XrayLogger");
const XraySummaryClient_1 = require("./XraySummaryClient");
const XrayDetailsClient_1 = require("./XrayDetailsClient");
const __1 = require("..");
class XrayClient {
    constructor(config) {
        const { serverUrl, logger = console, username, password, accessToken, proxy, headers, retries, } = config;
        if (!serverUrl) {
            throw new Error('Xray client : must provide platformUrl or xrayUrl');
        }
        this.httpClient = new HttpClient_1.HttpClient({ serverUrl, username, password, accessToken, proxy, headers, retries });
        this.logger = new XrayLogger_1.XrayLogger(logger);
    }
    summary() {
        return new XraySummaryClient_1.XraySummaryClient(this.httpClient, this.logger);
    }
    system() {
        return new XraySystemClient_1.XraySystemClient(this.httpClient, this.logger);
    }
    details() {
        return new XrayDetailsClient_1.XrayDetailsClient(this.httpClient, this.logger);
    }
    scan() {
        return new __1.XrayGraphClient(this.httpClient, this.logger);
    }
}
exports.XrayClient = XrayClient;
