// nuget-deps-tree/dist/src/Structure/OutputStructure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Project = exports.Tree = void 0;
class Tree {
    constructor(_projects) {
        this._projects = _projects;
    }
    get projects() {
        return this._projects;
    }
    set projects(value) {
        this._projects = value;
    }
    // Remove underscore when converting to JSON.
    toJSON() {
        return {
            projects: this._projects,
        };
    }
}
exports.Tree = Tree;
class Project {
    constructor(_name, _dependencies) {
        this._name = _name;
        this._dependencies = _dependencies;
    }
    get name() {
        return this._name;
    }
    set name(value) {
        this._name = value;
    }
    get dependencies() {
        return this._dependencies;
    }
    set dependencies(value) {
        this._dependencies = value;
    }
    // Remove underscore when converting to JSON.
    toJSON() {
        return {
            name: this._name,
            dependencies: this._dependencies,
        };
    }
}
exports.Project = Project;
