// jfrog-client-js/dist/src/Artifactory/ArtifactoryClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArtifactoryClient = void 0;
const HttpClient_1 = require("../HttpClient");
const ArtifactoryLogger_1 = require("./ArtifactoryLogger");
const ArtifactorySystemClient_1 = require("./ArtifactorySystemClient");
const ArtifactorySearchClient_1 = require("./ArtifactorySearchClient");
const ArtifactoryDownloadClient_1 = require("./ArtifactoryDownloadClient");
class ArtifactoryClient {
    constructor(config) {
        const { serverUrl, logger = console, username, password, accessToken, proxy, headers, retries, } = config;
        if (!serverUrl) {
            throw new Error('Artifactory client : must provide platformUrl or artifactoryUrl');
        }
        this.httpClient = new HttpClient_1.HttpClient({ serverUrl, username, password, accessToken, proxy, headers, retries });
        this.logger = new ArtifactoryLogger_1.ArtifactoryLogger(logger);
    }
    system() {
        return new ArtifactorySystemClient_1.ArtifactorySystemClient(this.httpClient, this.logger);
    }
    search() {
        return new ArtifactorySearchClient_1.ArtifactorySearchClient(this.httpClient, this.logger);
    }
    download() {
        return new ArtifactoryDownloadClient_1.ArtifactoryDownloadClient(this.httpClient, this.logger);
    }
}
exports.ArtifactoryClient = ArtifactoryClient;
