// jfrog-client-js/dist/src/Artifactory/ArtifactorySystemClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArtifactorySystemClient = void 0;
class ArtifactorySystemClient {
    constructor(httpClient, logger) {
        this.httpClient = httpClient;
        this.logger = logger;
        this.pingEndpoint = '/api/system/ping';
        this.versionEndpoint = '/api/system/version';
        this.usageEndpoint = '/api/system/usage';
    }
    async ping() {
        this.logger.debug('Sending ping request...');
        const requestParams = {
            url: this.pingEndpoint,
            method: 'GET',
            timeout: 2000,
        };
        try {
            return await this.httpClient.doRequest(requestParams);
        }
        catch (error) {
            return false;
        }
    }
    async version() {
        this.logger.debug('Sending version request...');
        const requestParams = {
            url: this.versionEndpoint,
            method: 'GET',
        };
        return await this.httpClient.doAuthRequest(requestParams);
    }
    async reportUsage(userAgent, featureArray) {
        this.logger.debug('Sending usage report...');
        const usageData = {
            productId: userAgent,
            features: featureArray,
        };
        const requestParams = {
            url: this.usageEndpoint,
            method: 'POST',
            data: usageData,
        };
        return await this.httpClient.doAuthRequest(requestParams);
    }
}
exports.ArtifactorySystemClient = ArtifactorySystemClient;
